src/v/cluster/cluster_epoch_generator.cc

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/cluster_epoch_generator.h"
+
+#include "base/vassert.h"
+#include "cluster/controller_stm.h"
+#include "cluster/logger.h"
+
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+
+namespace cluster {
+
+namespace {
+
+using namespace std::chrono_literals;
+
+// TODO(cloud-topics): make these configuration knobs
+constexpr ss::lowres_clock::duration epoch_cache_timeout = 1s;
+// The maximum amount of time to tolerate the same epoch being returned
+constexpr ss::lowres_clock::duration max_same_epoch_duration = 1min;
+
+// NOLINTNEXTLINE(*-avoid-reference-coroutine-parameters)
+ss::future<> force_increment_loop(controller_stm& stm, ss::abort_source& as) {
+    model::offset last_offset = stm.committed_offset();
+    while (!as.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable<ss::lowres_clock>(
+              max_same_epoch_duration, as);
+        } catch (const ss::sleep_aborted&) {
+            co_return;
+        }
+        model::offset new_offset = stm.committed_offset();
+        if (new_offset > last_offset) {
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
+          ss::lowres_clock::now() + 3s);
```

Comment:
In my local setup this causes a bunch of extra snapshots to be taken - should we make the snapshot mechanism ignore these empty batch writes?

Reply:
I'll add a JIRA for it. Do we always eagerly create a snapshot? I feel like they should only be every N bytes or N commits to the log, it seems better to replay a couple of updates than rewrite multi-MiB for every update (especially since large scale clusters/snapshots are more likely to get more udpates)